src/iop/rcd_demosaic.c

```diff
@@ -0,0 +1,621 @@
+              + 10.f * (cfa[indx - w3] * cfa[indx - w3] + cfa[indx + w3] * cfa[indx + w3])
+              + cfa[indx - w4] * cfa[indx - w4]
+              + cfa[indx + w4] * cfa[indx + w4]
+              );
+
+            const float cfai = cfa[indx];
+            const float H_Stat = fmaxf(epssq,
+                -18.f * cfai * (cfa[indx -  1] + cfa[indx +  1] + 2.0f * (cfa[indx -  2] + cfa[indx +  2]) - cfa[indx -  3] - cfa[indx +  3])
+                - 2.0f * cfai * (cfa[indx -  4] + cfa[indx +  4] - 19.f * cfai)
+                - cfa[indx -  1] * (70.f * cfa[indx +  1] + 12.f * (cfa[indx -  2] - cfa[indx -  4] - 2.0f * cfa[indx +  2]) + 38.f * cfa[indx -  3] - 16.f * cfa[indx +  3] + 6.f * cfa[indx +  4] - 46.f * cfa[indx -  1])
+                + cfa[indx +  1] * (24.f * cfa[indx -  2] + 12.f * (cfa[indx +  4] - cfa[indx +  2]) + 16.f * cfa[indx -  3] - 38.f * cfa[indx +  3] -  6.f * cfa[indx -  4] + 46.f * cfa[indx +  1])
+                + cfa[indx -  2] * (14.f * cfa[indx +  2] - 12.f * cfa[indx +  3] - 2.0f * cfa[indx -  4] + 2.0f * cfa[indx +  4] + 11.f * cfa[indx -  2])
+                + cfa[indx +  2] * (-12.f * cfa[indx -  3] + 2.0f * (cfa[indx -  4] - cfa[indx +  4]) + 11.f * cfa[indx +  2])
+                + cfa[indx -  3] * (2.0f * cfa[indx +  3] - 6.f * cfa[indx -  4] + 10.f * cfa[indx -  3])
+                + cfa[indx +  3] * (-6.f * cfa[indx +  4] + 10.f * cfa[indx +  3])
+                + cfa[indx -  4] * cfa[indx -  4]
+                + cfa[indx +  4] * cfa[indx +  4]);
+            VH_Dir[indx] = V_Stat / (V_Stat + H_Stat);
+          }
+        }
+
+        // STEP 2: Calculate the low pass filter
+        // Step 2.1: Low pass filter incorporating green, red and blue local samples from the raw data
+        // as an index>>1 access breaks proper vectorizing we use an extra index for the lpf results
+        for(int row = 2; row < tileRows - 2; row++)
+        {
+          for(int col = 2 + (FCRCD(row, 0) & 1), indx = row * RCD_TILESIZE + col, lp_indx = indx / 2; col < tileCols - 2; col += 2, indx +=2, lp_indx++)
+          {
+            lpf[lp_indx] = 0.25f * cfa[indx]
+                        + 0.125f * (cfa[indx - w1] + cfa[indx + w1] + cfa[indx - 1] + cfa[indx + 1])
+                       + 0.0625f * (cfa[indx - w1 - 1] + cfa[indx - w1 + 1] + cfa[indx + w1 - 1] + cfa[indx + w1 + 1]);
+          }
+        }
+
+        // STEP 3: Populate the green channel
+        // Step 3.1: Populate the green channel at blue and red CFA positions
+        for(int row = 4; row < tileRows - 4; row++)
+        {
+          int col = 4 + (FCRCD(row, 0) & 1);
+          int indx = row * RCD_TILESIZE + col;
+          int lp_indx = indx / 2;
+          // There has been quite some performance testing for the generic vs. SSE2 optimized loop, as the generated generic code (-O3)
+          // is not as fast as Ingos optmized code (~4% loss) we keep the SSE2 specific path.
+#ifdef __SSE2__
```

Comment:
can you explain why there is a #ifdef there so some code with having SSE2 and no else part ? That is, there is something done only when SSE2 is supported ? Looks suspicious to me... At least, this certainly deserve a comment.

Reply:
Because the loop counter is initialized before the loop and not inside the loop header, the code after the `#endif` handles the remaining pixels of the SSE code (if there are any) as well as the non SSE code. That's a pattern which reduces code duplication and in RawTherapee we get this question at least once per year ;-)

Reply:
So as I said, we need a comment :)

Reply:
I couldn't have this explained better, was just committing the compiler issue :-)